src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,346 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+//#define ASSERT_BALANCE
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+using Microsoft.Build.Eventing;
+
+namespace Microsoft.Build.Framework
+{
+    /// <summary>
+    /// A StringBuilder lookalike that reuses its internal storage.
+    /// </summary>
+    /// <remarks>
+    /// This class is being deprecated in favor of SpanBasedStringBuilder in StringTools. Avoid adding more uses.
+    /// </remarks>
+    internal sealed class ReuseableStringBuilder : IDisposable
+    {
+        /// <summary>
+        /// Captured string builder.
+        /// </summary>
+        private StringBuilder _borrowedBuilder;
+
+        /// <summary>
+        /// Capacity of borrowed string builder at the time of borrowing.
+        /// </summary>
+        private int _borrowedWithCapacity;
+
+        /// <summary>
+        /// Capacity to initialize the builder with.
+        /// </summary>
+        private int _capacity;
+
+        /// <summary>
+        /// Create a new builder, under the covers wrapping a reused one.
+        /// </summary>
+        internal ReuseableStringBuilder(int capacity = 16) // StringBuilder default is 16
+        {
+            _capacity = capacity;
+
+            // lazy initialization of the builder
+        }
+
+        /// <summary>
+        /// The length of the target.
+        /// </summary>
+        public int Length
+        {
+            get { return (_borrowedBuilder == null) ? 0 : _borrowedBuilder.Length; }
+            set
+            {
+                LazyPrepare();
+                _borrowedBuilder.Length = value;
+            }
+        }
+
+        /// <summary>
+        /// Convert to a string.
+        /// </summary>
+        public override string ToString()
+        {
+            if (_borrowedBuilder == null)
+            {
+                return String.Empty;
+            }
+
+            return _borrowedBuilder.ToString();
+        }
+
+        /// <summary>
+        /// Dispose, indicating you are done with this builder.
+        /// </summary>
+        void IDisposable.Dispose()
```

Comment:
Making this explicit makes it easier to miss disposing instances of this type, as the `Dispose` method doesn't appear in IntelliSense.

Reply:
done